packages/two-factor-connector/src/modules/sagas.js

```diff
@@ -9,35 +9,73 @@ const twoFactorActiveState = '1'
 
 export const usernameSelector = state => state.twoFactorConnector.username
 export const secretSelector = state => state.twoFactorConnector.secret
+export const inputSelector = state => state.input
 
 export function* requestSecret() {
-  const username = yield select(usernameSelector)
-  const principalsResponse = yield call(rest.requestSaga, `principals/${username}/two-factor`, {method: 'GET'})
-  const {secret, totpUri} = principalsResponse.body
-  yield put(actions.setSecret({text: secret, uri: totpUri}))
+  if (!(yield select(secretSelector))) {
+    const username = yield select(usernameSelector)
+    const principalsResponse = yield call(rest.requestSaga, `principals/${username}/two-factor`, {method: 'GET'})
+    const {secret, totpUri} = principalsResponse.body
+    yield put(actions.setSecret({secret, uri: totpUri}))
+  }
   yield put(actions.goToSecret())
 }
 
-export function* verifyCode({payload: {userCode}}) {
-  const username = yield select(usernameSelector)
-  const {text: secret} = yield select(secretSelector)
+function* loginAndActivateTwoFactor(secret, userCode, username, password) {
+  const data = {
+    secret,
+    userCode,
+    username,
+    password
+  }
+  const response = yield call(rest.requestSaga, `${__BACKEND_URL__}/nice2/login`,
+    {
+      method: 'POST',
+      headers: new Headers({
+        'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8'
+      }),
+      body: Object.keys(data)
+        .filter(k => !!data[k])
+        .sort()
+        .map(k => `${encodeURIComponent(k)}=${encodeURIComponent(data[k])}`).join('&')
+    })
+
+  yield put(actions.setSetupSuccessful(response.TWOSTEPLOGIN_ACTIVATION.success))
+}
 
+function* activateTwoFactor(username, secret, code) {
   const codeVerifyResponse = yield call(rest.requestSaga, `principals/${username}/two-factor`,
     {
       method: 'POST',
       body: {
         secret,
-        code: userCode
+        code
       },
       acceptedStatusCodes: [400]
     })
 
   yield put(actions.setSetupSuccessful(codeVerifyResponse.status === 204))
+}
+
+export function* verifyCode({payload: {userCode}}) {
+  const username = yield select(usernameSelector)
+  const {secret} = yield select(secretSelector)
+  const {password, forced} = yield select(inputSelector)
+
+  if (forced) {
+    yield call(loginAndActivateTwoFactor, secret, userCode, username, password)
```

Comment:
Would be nice to verify this code with the same endpoint. If not possible a rename of this function and maybe a comment in the code will be helpful.